### PR TITLE
feat: Implement new Symbolication Metrics

### PR DESCRIPTION
**Pull request by Swatinem**
This collects metrics on the number of scanned or unsymbolicated frames,
those referencing unmapped modules, and stack traces that do not end in
well known thread bases or contain any of the mentioned bad frames.

### CODE REVIEW

File: crates/symbolicator/src/services/symbolication.rs [hunk 1]
```diff
@@ -1113,22 +1113,104 @@ fn symbolicate_frame(
...
+    /// These may be the result of issues in the client-side module finder, or
+    /// broken debug-id information.
+    ///
+    /// We can improve this by fixing client-side implementations and having
+    /// proper debug-ids.
+    unmapped_frames: u64,
+}
+
+impl StacktraceMetrics {
+    fn update(&mut self, other: StacktraceMetrics) {
```
Comment (loewenheim) [inline crates/symbolicator/src/services/symbolication.rs:1153]: This could work well as a `PlusAssign` impl.

**Review by loewenheim — APPROVED** [review]

File: crates/symbolicator/src/services/symbolication.rs (hunk 1 above)
Comment (Swatinem) [reply crates/symbolicator/src/services/symbolication.rs:1153]: I thought about it, but IMO this conveys the intent better.

Comment (loewenheim) [reply crates/symbolicator/src/services/symbolication.rs:1153]: Yeah, that's fair.

File: crates/symbolicator/src/services/symbolication.rs [hunk 2]
```diff
@@ -1113,22 +1113,104 @@ fn symbolicate_frame(
...
+
+    false
+}
+
 fn symbolicate_stacktrace(
     thread: RawStacktrace,
     caches: &SymCacheLookup,
     signal: Option<Signal>,
-) -> CompleteStacktrace {
+) -> (CompleteStacktrace, StacktraceMetrics) {
```
Comment (flub) [inline crates/symbolicator/src/services/symbolication.rs:1193]: why do we have to pass this back?  I didn't spot any usage that you couldn't record here directly and you wouldn't need to complicate these signatures.  You also don't really need to declare a new struct with a bunch of implicit but undeclared metric types and update semantics as you can just call `metric!` directly which immediately shows you the metric type in the call.

File: crates/symbolicator/src/services/symbolication.rs [hunk 3]
```diff
@@ -1265,11 +1372,24 @@ impl SymbolicationActor {
             let modules: Vec<_> = modules.into_iter().map(|(_, module)| module).collect();
 
             metric!(time_raw("symbolication.num_modules") = modules.len() as u64);
+            metric!(
+                time_raw("symbolication.unusable_modules") =
```
Comment (flub) [inline crates/symbolicator/src/services/symbolication.rs:1376]: We really could do with an alias that calls `time_raw` `histogram` instead...

File: crates/symbolicator/src/services/symbolication.rs [hunk 4]
```diff
@@ -1265,11 +1372,24 @@ impl SymbolicationActor {
...
+                    modules.iter().filter(|m| m.raw.debug_id.is_none()).count() as u64
+            );
+
             metric!(time_raw("symbolication.num_stacktraces") = stacktraces.len() as u64);
+            metric!(time_raw("symbolication.truncated_stacktraces") = metrics.truncated_traces);
+            metric!(time_raw("symbolication.bad_stacktraces") = metrics.bad_traces);
```
Comment (flub) [inline crates/symbolicator/src/services/symbolication.rs:1382]: i thought this was a counter?  like, each symbolication only sets this once

File: crates/symbolicator/src/services/symbolication.rs (hunk 3 above)
Comment (jan-auer) [reply crates/symbolicator/src/services/symbolication.rs:1376]: Agreed, we should align this with Relay's metrics macros, with the exception of the enums.

Comment (flub) [reply crates/symbolicator/src/services/symbolication.rs:1376]: hmm, maybe, maybe not.  let's not jump into that right now 😄

Comment (flub) [reply crates/symbolicator/src/services/symbolication.rs:1376]: sorry, being more transparent: i still think metrics-rs is the way to go.  creating a simple alias for a function right now i think is reasonable, spending more time on copying relay is likely not without evaluating things a bit more carefully.  It's not like metrics are a major pain point right now.

File: crates/symbolicator/src/services/symbolication.rs [hunk 5]
```diff
@@ -1113,9 +1113,78 @@ fn symbolicate_frame(
     Ok(rv)
 }
 
+/// Stacktrace related Metrics
+///
+/// This gives some metrics about the quality of the stack traces included
+/// in a symbolication request. See the individual members for more information.
```
Comment (flub) [inline crates/symbolicator/src/services/symbolication.rs:1119]: maybe say these are all supposed to be recorded as histograms.

**Review by flub — APPROVED** [review]

**Pull request merged**